### PR TITLE
override the version number that is read from pubspec.lock file for all native plugins when generating the local Maven repository.

### LINKED ISSUE #152538: The generated aar for native plugin in build/host/outputs/repo should have a correct version number.
### Steps to reproduce

1. create a flutter project
2. add a plugin with native code, such as image_picker(It depens on image_picker_android)
3. build AAR for android with `flutter  build aar --build-number 2.0`

### Expected results

The generated aar for image_picker_android in build/host/outputs/repo should have the correct version number.

### Actual results

It always has the same version of `--build-number`, even though I didn't change the version number of image_picker in pubspec.yaml.

### Code sample

<details open><summary>Code sample for pubspec.yaml</summary>

```yaml
version: 1.0.0+1

environment:
  sdk: ^3.6.0-91.0.dev

dependencies:
  flutter:
    sdk: flutter
  image_picker: 1.0.0

dev_dependencies:
  flutter_lints: ^4.0.0

flutter:
  module:
    androidX: true
    androidPackage: com.example.flutter_project
    iosBundleIdentifier: com.example.flutterProject
```

</details>

### Screenshots or Video

_No response_

### Logs

_No response_

### Flutter Doctor output

<details open><summary>Doctor output</summary>

```console
[✓] Flutter (Channel master, 3.24.0-1.0.pre.349, on macOS 14.5 23F79 darwin-x64, locale zh-Hans-CN)
[✓] Android toolchain - develop for Android devices (Android SDK version 35.0.0)
[!] Xcode - develop for iOS and macOS (Xcode 15.4)
    ✗ Unable to get list of installed Simulator runtimes.
[✓] Chrome - develop for the web
[✓] Android Studio (version 2024.1)
[✓] Connected device (3 available)
[✓] Network resources

```

</details>

### PR DESCRIPTION
**Pull request by chunfengyao**
Fixed https://github.com/flutter/flutter/issues/152538
override the version number that is read from pubspec.lock file for all native plugins when generating the local Maven repository.
This pull request addresses this issue by obtaining the version number from the pubspec.lock file, which contains the actual version numbers of plugins in pub.dev. When determining the version number that should be used in the AAR file, most plugin's defined versions are set to '1.0' in their respective `android/build.gradle` files. To avoid significant changes, this pull request selects the plugin's version published on Pub server.
I would like to express my gratitude to all reviewers who have contributed to this pull request.

There is no issue linked to this pull request. If necessary, I could create one.

This pull request has no changes in the [flutter/tests] repo. If a test is necessary, I could create one.

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] I followed the [breaking change policy] and added [Data Driven Fixes] where supported.
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#overview
[Tree Hygiene]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md
[test-exempt]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/blob/main/docs/contributing/Style-guide-for-Flutter-repo.md
[Features we expect every widget to implement]: https://github.com/flutter/flutter/blob/main/docs/contributing/Style-guide-for-Flutter-repo.md#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/blob/main/docs/contributing/Chat.md
[Data Driven Fixes]: https://github.com/flutter/flutter/blob/main/docs/contributing/Data-driven-Fixes.md

### CODE REVIEW

**Comment by chunfengyao** [comment]
> Thanks for your pull request! It looks like this may be your first contribution to a Google open source project. Before we can look at your pull request, you'll need to sign a Contributor License Agreement (CLA).
> 
> View this [failed invocation](https://github.com/flutter/flutter/pull/152467/checks?check_run_id=28031676377) of the CLA check for more information.
> 
> For the most up to date status, view the checks section at the bottom of the pull request.

I've changed my git mail to the correct one.

**Comment by chunfengyao** [comment]
@reidbaker  Could you please help review this pull request? Thanks!

File: packages/flutter_tools/gradle/aar_init_script.gradle [hunk 1]
```diff
@@ -8,7 +8,24 @@ import org.gradle.api.Project
...
+    }
+    dependencies {
+        classpath 'org.yaml:snakeyaml:2.2+'
+    }
+}
+
+
+ext{
+    pubspecLock = new org.yaml.snakeyaml.Yaml().load(
+            new File(new File(System.getProperty('user.dir')).parentFile, "pubspec.lock").newInputStream()
```
Comment (chunfengyao) [inline packages/flutter_tools/gradle/aar_init_script.gradle:24]: Defining the YAML instance here could reduce the number of file reads from disk caused by multiple calls to 'configureProject'.

File: packages/flutter_tools/gradle/aar_init_script.gradle [hunk 2]
```diff
@@ -22,7 +39,17 @@ void configureProject(Project project, String outputDir) {
     // to resolve dependencies.
     project.version = project.version.replace("-SNAPSHOT", "")
 
-    if (project.hasProperty("buildNumber")) {
+    if (!isModuleProject) {
+        def deps = pubspecLock.packages.find{ it.key == "${project.name}" }
```
Comment (chunfengyao) [inline packages/flutter_tools/gradle/aar_init_script.gradle:43]: Matching plugin name from the YAML instance.

File: packages/flutter_tools/gradle/aar_init_script.gradle [hunk 3]
```diff
@@ -22,7 +39,17 @@ void configureProject(Project project, String outputDir) {
...
+        if (deps != null && deps.value != null && deps.value.version != null && !(deps.value.version.trim().isEmpty())){
+            def _version = deps.value.version
+            project.version = _version
```
Comment (chunfengyao) [inline packages/flutter_tools/gradle/aar_init_script.gradle:46]: Override default version number with the version number obtained from the YAML instance.

File: packages/flutter_tools/gradle/aar_init_script.gradle [hunk 4]
```diff
@@ -76,7 +103,7 @@ void configurePlugin(Project project, String outputDir) {
         // A plugin doesn't support the Android platform when this property isn't defined in the plugin.
         return
     }
-    configureProject(project, outputDir)
+    configureProject(project, outputDir, false)
```
Comment (chunfengyao) [inline packages/flutter_tools/gradle/aar_init_script.gradle:106]: For module projects, it should use '--builder-number' as the version number; therefore, I passed the project type to the 'configureProject' function.

File: packages/flutter_tools/gradle/aar_init_script.gradle [hunk 5]
```diff
@@ -121,14 +148,14 @@ projectsEvaluated {
     if (rootProject.property("is-plugin").toBoolean()) {
         assert rootProject.hasProperty("output-dir")
         // In plugin projects, the root project is the plugin.
-        configureProject(rootProject, rootProject.property("output-dir"))
+        configureProject(rootProject, rootProject.property("output-dir"), false)
```
Comment (chunfengyao) [inline packages/flutter_tools/gradle/aar_init_script.gradle:151]: For module projects, it should use '--builder-number' as the version number; therefore, I passed the project type to the 'configureProject' function.

File: packages/flutter_tools/gradle/aar_init_script.gradle [hunk 6]
```diff
@@ -121,14 +148,14 @@ projectsEvaluated {
...
         return
     }
     // The module project is the `:flutter` subproject.
     Project moduleProject = rootProject.subprojects.find { it.name == "flutter" }
     assert moduleProject != null
     assert moduleProject.hasProperty("output-dir")
-    configureProject(moduleProject, moduleProject.property("output-dir"))
+    configureProject(moduleProject, moduleProject.property("output-dir"), true)
```
Comment (chunfengyao) [inline packages/flutter_tools/gradle/aar_init_script.gradle:158]: Same as [here](https://github.com/flutter/flutter/pull/152467/files#r1696807347).

File: packages/flutter_tools/test/integration.shard/android_plugin_version_number.dart [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+
+    final String projectPath = fileSystem.path.join(tempDir.path, 'flutter_project');
+
+    final File modulePubspec = fileSystem.file(fileSystem.path.join(projectPath, 'pubspec.yaml'));
+    String pubspecContent = modulePubspec.readAsStringSync();
+    pubspecContent = pubspecContent.replaceFirst(
+      'dependencies:',
+      '''
+dependencies:
+  image_picker: ^1.0.0''',
```
Comment (chunfengyao) [inline packages/flutter_tools/test/integration.shard/android_plugin_version_number.dart:52]: Perhaps using 'image_picker_android' directly here would be better.

**Comment by reidbaker** [comment]
I am trying to get a better understanding of when pubspec.lock files are expected to change.

**Comment by chunfengyao** [comment]
> I am trying to get a better understanding of when pubspec.lock files are expected to change.

The rule of package management for Pub dictates that the `pubspec.lock` file stores the version numbers of every library used. Therefore, if I change the version number in `pubspec.yaml` and then run flutter pub get, or upgrade all upgradable packages with flutter pub update, it may affect the version constraints in `pubspec.lock`. Eventually, this may change the content of `pubspec.lock` file.

File: packages/flutter_tools/gradle/aar_init_script.gradle [hunk 8]
```diff
@@ -8,7 +8,24 @@ import org.gradle.api.Project
...
 import org.gradle.api.publish.maven.MavenPublication
 
-void configureProject(Project project, String outputDir) {
+initscript {
+    repositories {
+        google()
+        mavenCentral()
+    }
+    dependencies {
+        classpath 'org.yaml:snakeyaml:2.2+'
```
Comment (reidbaker) [inline packages/flutter_tools/gradle/aar_init_script.gradle:17]: We probably do not want to add a new gradle library dependency.

**Review by reidbaker — CHANGES_REQUESTED** [review]
This pull request and the issue are a bit hard to follow. I am going to try to reproduce the issue on my machine to see if that helps me understand the issue better. 

In general we do not like to add dependencies we do no control. Especially for code used to build assets our developers use. 

Also passing in a boolean for "isModuleProject" implies more than is needed for configureProject. I suggest renaming the booling to something more tightly scoped to what happens when the boolean is true/false. 

Finally this pull request, as written, has confused multiple contributors and I get the sense that the pr description and the bug were written in a hurry. To help this pr land assume the reviewers have a broad knowledge of flutter but have less experience in this specific area when communicating what is changing or what needs to be changed.

**Comment by chunfengyao** [comment]
> This pull request and the issue are a bit hard to follow. I am going to try to reproduce the issue on my machine to see if that helps me understand the issue better.
> 
> In general we do not like to add dependencies we do no control. Especially for code used to build assets our developers use.
Right, perhaps I could use a simplified function to process YAML files instead of this gradle library.

> 
> Also passing in a boolean for "isModuleProject" implies more than is needed for configureProject. I suggest renaming the booling to something more tightly scoped to what happens when the boolean is true/false.
Should I use the project name as a parameter for the function "configureProject"? Something like :flutter or :app?

> 
> Finally this pull request, as written, has confused multiple contributors and I get the sense that the pr description and the bug were written in a hurry. To help this pr land assume the reviewers have a broad knowledge of flutter but have less experience in this specific area when communicating what is changing or what needs to be changed.
Got it, I'll add more descriptive details to the issue and PR descriptions.

File: packages/flutter_tools/gradle/aar_init_script.gradle (hunk 8 above)
Comment (chunfengyao) [reply packages/flutter_tools/gradle/aar_init_script.gradle:17]: It's easy to parse YAML using this library, but parsing YAML manually is difficult. If library usage isn't feasible, I'll attempt manual parsing.

Comment (reidbaker) [reply packages/flutter_tools/gradle/aar_init_script.gradle:17]: Hold off on that for now. Let us review the idea of the pr and then we can take a look at yaml parsing. I dont want you to write manual yaml parsing if this cl has other things that prevent us from merging.

**Comment by andrewkolos** [comment]
@reidbaker Is the last comment here enough to move forward with? Or, is this PR stuck and should we consider closing this?

**Comment by chunfengyao** [comment]
> > I believe most plugin maintainers should and will update the plugin's version number when they modify the code and re-publish it.
> 
> It's impossible to re-publish the same version; I'm talking about cases like editing the source of a `path` based plugin, or using an unpinned `git` source for a plugin, where the source can easily change without the version changing.
> 
> > Even for locally sourced plugins, running `flutter build aar` triggers the gradle `assemble` task, which always uses the latest code to generate the aar file.
> 
> But then under your proposed change, you would have different build output with the same version number, which directly violates the purpose of `build-number`.


I don't have a good solution for this https://github.com/flutter/flutter/issues/152538#issuecomment-2265075353 at the moment. Personally, I think if users are including dependencies via a git path or a local path, they should be responsible for updating the uploaded aar file after each build. Alternatively, perhaps we could add a build flag to 'flutter build aar' that forces all dependencies to use the version number specified by the '--build-number' flag?

**Comment by chunfengyao** [comment]
Alternatively, should we include more details in the generated `pubspec.lock` file, such as dependency type? This would allow us to apply the version number passed from `--build-number` to both git based and path based dependencies.

**Comment by andrewkolos** [comment]
Hello. Unfortunately, it looks like we aren't able to make progress with this PR due to its complexity and available reviewer bandwidth. Thank you for your contribution.
